src/components/GlobalLeaderboards/Context.tsx

```diff
@@ -0,0 +1,213 @@
+      setPositionsOrderBy(key);
+      setPositionsOrderDirection(1);
+    }
+  }, [positionsOrderBy]);
+
+  const nextPositionsHash = utils.sha256(topPositionsUnordered.data ? topPositionsUnordered.data.map(a => (
+    a[positionsOrderBy]!.toString().split("").map((c: string) => c.charCodeAt(0))
+  )).flat() : []);
+
+  if (positionsHash !== nextPositionsHash) {
+    setPositionsHash(nextPositionsHash);
+  }
+
+  useEffect(() => {
+    if (!topPositionsUnordered.data) {
+      return;
+    }
+
+    const data = [...topPositionsUnordered.data].sort((a, b) => {
+      const key = positionsOrderBy;
+      if (a[key] instanceof BigNumber && b[key] instanceof BigNumber) {
+        return positionsOrderDirection * ((a[key] as BigNumber).gt(b[key] as BigNumber) ? -1 : 1);
+      } else {
+        // TODO: remove log
+        // eslint-disable-next-line no-console
+        console.warn("Not a BigNumber", { key, a, b });
+        return 1;
+      }
+    }).map((p, i) => ({ ...p, rank: i }));
+
+    setTopPositions(data);
+    if (p.current && !p.current.pSorted) {
+      p.current("sorted top positions");
+      p.current.pSorted = true;
+    }
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [positionsHash, positionsOrderBy, positionsOrderDirection]);
+
+  const nextAccountsHash = utils.sha256(topAccountsUnordered.data ? topAccountsUnordered.data.map(a => (
+    a[accountsOrderBy]!.toString().split("").map((c: string) => c.charCodeAt(0))
+  )).flat() : []);
+
+  if (accountsHash !== nextAccountsHash) {
+    setAccountHash(nextAccountsHash);
+  }
+
+  useEffect(() => {
+    if (!topAccountsUnordered.data) {
+      return;
+    }
+
+    const data = [...topAccountsUnordered.data];
+    data.sort((a, b) => {
+      const key = accountsOrderBy;
+      if (a[key] instanceof BigNumber && b[key] instanceof BigNumber) {
+        return accountsOrderDirection * ((a[key] as BigNumber).gt(b[key] as BigNumber) ? -1 : 1);
+      } else {
+        // TODO: remove log
+        // eslint-disable-next-line no-console
+        console.warn("Not a BigNumber", {key, a, b});
+        return 1;
+      }
+    });
+
+    data.forEach((a, i) => a.rank = i);
+
+    setTopAccounts(data);
+    if (p.current && !p.current.aSorted) {
+      p.current("sorted top accounts");
+      p.current.aSorted = true;
+    }
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [accountsHash, accountsOrderBy, accountsOrderDirection]);
+
+  if (topAccounts && topAccounts.length && !p.current.aReturned) {
+    p.current("returning top accounts");
+    p.current.aReturned = true;
+  }
+
+  if (topPositions && topPositions.length && !p.current.pReturned) {
+    p.current("returning top accounts");
+    p.current.pReturned = true;
+  }
+
+  if (topAccounts && topAccounts.length && topPositions && topPositions.length &&!p.current.reported) {
+    p.current("returning all data");
+    p.current.report();
+    p.current.reported = true;
+  }
+  const context = {
```

Comment:
i this case all context consumers will re-render every time at every Provider state update

i think it more safer to wrap it into useMemo

Reply:
AFAIU it will only re-render when React UI tree nodes change. In this case it only happens when the relevant data changes.

Again, please correct me if I am wrong.

Reply:
just tested in Code sandbox. 

it re-renders when "value" object is changed, in this case at every render of the context

React Context doesn't provide some smart way to pick only neccesary data from state as Redux does

Reply:
@divhead, what do you mean by re-render? If you mean expensive DOM manipulations, they only occur when the html tree changes, which is not our case as we always render the same table with the same elements in all of its' rows. I believe changes in the "value" object, even if there are many of those, are cheap as long as they do not trigger significant DOM changes. In our case they only change html attributes that correspond to the currently displayed data. AFAIK this is a React feature that is available irregardless of whether we use Context or not.

Reply:
discussed in DM